integration/test/Testlib/App.hs

```diff
@@ -0,0 +1,935 @@
+      MonadCatch,
+      MonadThrow,
+      MonadReader Env
+    )
+
+data AppFailure = AppFailure String
+
+instance Show AppFailure where
+  show (AppFailure msg) = msg
+
+instance Exception AppFailure where
+  displayException (AppFailure msg) = msg
+
+failApp :: String -> App a
+failApp msg = throw (AppFailure msg)
+
+runAppWithEnv :: Env -> App a -> IO a
+runAppWithEnv e m = runReaderT (unApp m) e
+
+getPrekey :: App Value
+getPrekey = App $ do
+  pks <- asks (.prekeys)
+  (i, pk) <- atomicModifyIORef pks getPK
+  pure $ object ["id" .= i, "key" .= pk]
+  where
+    getPK [] = error "Out of prekeys"
+    getPK (k : ks) = (ks, k)
+
+getLastPrekey :: App Value
+getLastPrekey = App $ do
+  pks <- asks (.lastPrekeys)
+  lpk <- atomicModifyIORef pks getPK
+  pure $ object ["id" .= lastPrekeyId, "key" .= lpk]
+  where
+    getPK [] = error "Out of prekeys"
+    getPK (k : ks) = (ks, k)
+
+    lastPrekeyId :: Int
+    lastPrekeyId = 65535
+
+instance IsTest (App ()) where
+  run opts action _ = do
+    env <- mkEnv (lookupOption opts)
+    result :: Tasty.Result <-
+      (runAppWithEnv env action >> pure (Tasty.testPassed ""))
+        `E.catches` [ E.Handler
+                        ( \(e :: AssertionFailure) -> do
+                            pure (testFailedDetails (displayException e) (printFailureDetails e))
+                        ),
+                      E.Handler
+                        (\(ex :: SomeException) -> pure (testFailed (show ex)))
+                    ]
+    pure result
+
+  testOptions = Tagged [Option (Proxy @ConfigFile)]
+
+data Env = Env
+  { context :: Context,
+    manager :: HTTP.Manager,
+    prekeys :: IORef [(Int, String)],
+    lastPrekeys :: IORef [String],
+    serviceConfigsDir :: FilePath,
+    servicesCwdBase :: Maybe FilePath
+  }
+
+data Context = Context
+  { serviceMap :: ServiceMap,
+    version :: Int
+  }
+
+data ServiceMap = ServiceMap
+  { brig :: HostPort,
+    galley :: HostPort,
+    cannon :: HostPort
+  }
+  deriving (Show, Generic)
+
+instance FromJSON ServiceMap
+
+data HostPort = HostPort
+  { host :: String,
+    port :: Word16
+  }
+  deriving (Show, Generic)
+
+instance FromJSON HostPort
+
+data Service = Brig | Galley | Cannon
+  deriving (Show, Eq, Ord)
+
+serviceName :: Service -> String
+serviceName srv = map C.toLower (show srv)
+
+serviceHostPort :: ServiceMap -> Service -> HostPort
+serviceHostPort m Brig = m.brig
+serviceHostPort m Galley = m.galley
+serviceHostPort m Cannon = m.cannon
+
+newtype ConfigFile = ConfigFile {unConfigFile :: FilePath}
+
+instance IsOption ConfigFile where
+  defaultValue = ConfigFile "services/integration.yaml"
+  parseValue = Just . ConfigFile
+  optionName = fromString "config"
+  optionHelp = fromString "Configuration file for integration tests. Default: services/integration.yaml"
+
+instance IsOption ServiceMap where
+  defaultValue = error "No Default value"
+  parseValue = const Nothing
+  optionName = fromString "config-loaded-from-file"
```

Comment:
?

Reply:
Ah, that instance is a leftover. I remove it